llama.cpp

```diff
@@ -5865,43 +5274,343 @@ static struct ggml_cgraph * llama_build_graph(
+            { "KQV_merged",                 offload_func_v  },
+            { "KQV_merged_contiguous",      offload_func_v  },
+
+            { "result_wo",                  offload_func    },
+            { "result_wo_b",                offload_func    },
+            { "inpL_+_result_wo",           offload_func    },
+
+            { "inpFF",                      offload_func    },
+
+            { "rms_norm_1",                 offload_func    },
+            { "ffn_norm",                   offload_func    },
+            { "ffn_norm_0",                 offload_func    },
+            { "ffn_norm_0_w",               offload_func    },
+            { "ffn_norm_0_wb",              offload_func    },
+
+            { "result_w3",                  offload_func    },
+            { "result_w3_b",                offload_func    },
+            { "result_w2",                  offload_func    },
+            { "result_w2_b",                offload_func    },
+            { "result_w1",                  offload_func    },
+
+            { "silu",                       offload_func    },
+            { "gelu",                       offload_func    },
+            { "relu",                       offload_func    },
+            { "sqr(relu)",                  offload_func    },
+
+            { "silu_x_result_w3",           offload_func    },
+            { "inpFF_+_result_w2",          offload_func    },
+            { "inpL_+_inpFF_+_result_w2",   offload_func    },
+
+            { "rms_norm_2",                 offload_func_nr },
+            { "out_norm_0",                 offload_func_nr },
+            { "out_norm_0_w",               offload_func_nr },
+
+          //{ "result_norm",                offload_func_nr }, // TODO CPU + GPU mirrored backend
+          //{ "result_output",              offload_func    },
+        };
+
+        static const std::unordered_map<offload_func_t, std::string> k_offload_func_name = {
+            { ggml_offload_nop,                  "CPU" },
+#ifdef GGML_USE_CUBLAS
+            { ggml_cuda_assign_buffers_no_alloc, "GPU (CUDA)" },
+#endif
+        };
+
+        std::unordered_map<std::string, int> ofn;
+
+        for (int i = 0; i < result->n_nodes; ++i) {
+            struct ggml_tensor * cur = result->nodes[i];
+
+            const std::string name = cur->name;
+
+            if (k_offload_func.find(name) == k_offload_func.end()) {
+                // if a tensor that is not view hasn't been offloaded, we warn the user
+                if (worst_case && cur->view_src == nullptr) {
+                    LLAMA_LOG_WARN("%s: node %4d %32s: not offloaded (ref: %s)\n", __func__,
+                            i, name.c_str(), "https://github.com/ggerganov/llama.cpp/pull/3837");
+                }
+
+                continue;
+            }
+
+            // count the number of layers and respect the provided n_gpu_layers
+            offload_func_t f = k_offload_func.at(name);
```

Comment:
Consider keeping the iterator returned by `k_offload_func.find(name)` to avoid a second lookup.

Reply:
With the latest change of passing a callback to the build functions, the performance is the same as on `master`.